### ISSUE TITLE
Fast preview generation for JPEG files via provider or app

### ISSUE THREAD

**Issue by fungs** [owncloud/core#17916]
Hi there,

this is related to all the issues with the thumbnail generation being slow but it is a fast solution for JPEG files in addition to any queuing or background thumbnail generation. I'm running ownCloud (8.1) on a server with fairly limited hardware (VIA Eden Processor 1500MHz, 500 MiB RAM). The image preview and gallery features are basically unusable for more than a handful of pictures. When looking for faster alternatives, I found that some programs which are specifically designed for JPEG-type data run the thumbnail generation much much faster.

E.g., for a single 23 MiB photo, if I run imagemagick (which should be comparable to GD), I get

```
  real  2m8.884s
  user  0m36.496s
  sys   0m6.880s
```

If I run the conversion with epeg, I get

```
  real  0m3.288s
  user  0m2.300s
  sys   0m0.160s
```

This is a huge improvement and it should be relatively easy to reimplement lib/private/preview/jpeg.php to use epeg/libepeg or a similar software and the benefit for users running OC on thin hardware will be large.

Also see
https://github.com/mattes/epeg
http://superuser.com/questions/249646/fastest-jpeg-thumbnail-generator-for-linux

**Comment by neiljp** [comment]
Is this definitely the slow element in the gallery? The preview/gallery is slow for me too, and I would appreciate any attempt being made to improve this, but is the problem with file loading, resizing, ...? The gallery view doesn't seem to be influenced so much by the size of the image (though I don't have 23MiB images!), but rather the number of images in the gallery?

**Comment by oparoz** [comment]
@fungs Since it requires external libraries, this is a good candidate for an app.
I would post on social media, forums and freelancing sites if you need this feature.

@neiljp Regarding Gallery/Pictures issues, on 8.1, you only notice a slowdown the first time you need to generate thumbnails and if you're still using Pictures.
Install Gallery from the multimedia section of your app manager and you should be able to navigate through your large pictures collection with relative ease.

I think this should be closed, unless ownCloud thinks this feature is critical to the success of ownCloud and should be in core. @DeepDiver1975 @LukasReschke @karlitschek

**Comment by fungs** [comment]
@aparoz Question one would be if it is possible to put core functionality like the image preview in the files view into an app. I would totally go for it.

I'm not only talking about the gallery app which I'm not usually using but about the instantanious preview generation in the files view which takes down the entire ownCloud instance because my apache process fills the single CPU with preview jobs instead of delivering the interface.

**Comment by oparoz** [comment]
@fungs - Yes, from 8.1 you can completely replace the JPEG provider with your own. That's the main reason there won't be any new provider being added to ownCloud.

**Comment by oparoz** [comment]
And you make a good point regarding the way previews are loaded in Files. It might be worth opening a separate issue for that since, on modern browsers, it's possible to use an alternative way to send previews back, using EventSource.

**Comment by neiljp** [comment]
@oparoz I noticed that (after some tweaking? I didn't take notes) the gallery did cache the 'page' of images, which was better, but it still takes ~5s per 'line' of images the first time through, which is quite slow! Other than that, the speed of either seemed to be comparable, from what I remember (pictures vs gallery), so maybe something else is at fault in my setup/server.

**Comment by oparoz** [comment]
@neiljp - One reason could be that you're browsing files stored on slow storage (external). Try the demo to see if it's as slow as what you get on your server.

**Comment by fungs** [comment]
Given that the typical majority of images are (still) in JPEG format and that there are many OC instances running on embedded or otherwise limited hardware, an improvement here would be much appreciated by those users. The discussion showed that the implementation of the JPEG preview provider in an app using epeg/libepeg would be possible. I'm not a PHP programmer but I would fully support someone who would be willing to write such functionality and can provide testing and benchmarking.

**Comment by oparoz** [comment]
There are still several hurdles

* It's not easy to install for non-sysadmins
* There are no PHP bindings, meaning you need to open your system by letting PHP use dangerous methods
* Feedback from a user: "...the quality leaves a lot to be desired, as it gets kind of blocky"

>  I'm not a PHP programmer but I would fully support someone who would be willing to write such functionality and can provide testing and benchmarking.

It depends on what you mean by fully support. It would probably cost several hundred $ to write such an app, including automated tests and documentation.

The alternative is of course to simply generate thumbnails in advance, using the occ command.

**Comment by fungs** [comment]
@oparoz did you notice any quality issues because I didn't (maybe because I did not look closely enough)?

You might be right that packaging and PHP language binding should be done first. Another solution, probably the best for OC, would be to get fast JPEG rescaling into the existing image manipulation libraries. I don't know how difficult that would be. Since there are a couple of functions which call external programs like pwauth already, I though this would be a possible option to add.

By supporting I meant as a programmer and IT technician. Maybe NAS vendors can provide financial support if they are interested in running OC on their hardware.

If I knew how to call an external program for preview generation in the cache on the command line, I would do that in a cron job.

**Comment by oparoz** [comment]
@fungs - I did not as I haven't tested the library. I just stumbled upon some comments when doing my research.

> Another solution, probably the best for OC, would be to get fast JPEG rescaling into the existing image manipulation libraries

oC is using the only readily available tool: GD. Any other library would have external dependencies.

>  Since there are a couple of functions which call external programs like pwauth already

I could not find any calls to pwauth in oC.
And it's easy to call a tool like ejeg to resize images, but it requires using methods which are often blacklisted for security reasons.

> By supporting I meant as a programmer and IT technician. Maybe NAS vendors can provide financial support if they are interested in running OC on their hardware.

Thanks for clarifying. NAS vendors would be good candidates since they provide the whole OS. They can offer the lib and harden the code.

> If I knew how to call an external program for preview generation in the cache on the command line, I would do that in a cron job.

You could try to modify the occ command which generate thumbnails (see above) and call epeg instead of using internal thumbnail generating methods. Another solution would be to raise funds to get that implemented. I'm sure many people with thousands of images will be interested if they can cut down the time it takes to generate thumbnails for their large collection, if the quality is as good as what they have now.
